Release.md

```diff
@@ -1,23 +1,49 @@
-# Releasing through Travis-CI
+# Release Instructions
 
-### Workflow 1
+## Dependencies
 
-1. Draft a new release on Github. Set the "Tag version" to the value of version in your application package.json, and prefix it with v. "Release title" can be anything you want.
-For example, if your application package.json version is 1.0, your draft's "Tag version" would be v1.0.
-2. Push some commits. Every CI build will update the artifacts attached to this draft.
-3. Once you are done, publish the release. GitHub will tag the latest commit for you.
+JupyterLab App uses [tbump](https://github.com/dmerejkowsky/tbump) to bump JupyterLab and app versions. You can install using:
+```bash
+pip install tbump
+```
 
-The benefit of this workflow is that it allows you to always have the latest artifacts, and the release can be published once it is ready.
+## Versioning
 
-### Workflow 2
+JupyterLab App needs to be versioned with the same major, minor and patch versions as the JupyterLab it bundles. For example, if JupyterLab App is based on JupyterLab 3.1.12, a valid JupyterLab App version is 3.1.12-1 to 3.1.12-n. Last number after `-` is used as the `build number`. This version matching is enforced before JupyterLab App installer binaries are published.
 
-1. run `npm version major/minor/patch` to tag and generate a commit
-2. push the commit and the tag
-3. The CI will draft a release with the correct "Tag version" from your package.json file
+JupyterLab version, that JupyterLab App bundles, is determined by `@jupyterlab/metapackage` dependency version in the [yarn.lock](yarn.lock).
 
+If the JupyterLab version is not changing with the new JupyterLab App release then only increment the build number after `-` (for example `3.1.12-2` to `3.1.12-3`). However, if JupyterLab version is changing with the new JupyterLab App release then reset the build number after `-` to 1 (for example `3.1.12-3` to `3.1.13-1`).
 
-# Dependencies
 
-To build for Linux and Windows the suggested way is to use the `dockerdist:linux` and `dockerdist:windows` command that use a well-tested and fully equipped docker environment to build for both.
-To build the dmg container for macOS, a macOS machine is required. It's suggested to use Travis-CI in this regard.
-This [documentation](https://github.com/electron-userland/electron-builder/wiki/Multi-Platform-Build) explains the needed dependencies to build locally.
+## Updating the bundled JupyterLab
+
+In order to change the JupyterLab version bundled with the App:
+
+1. Update all the dependencies in [package.json](package.json) (packages with `@jupyterlab` scope and also `@jupyter-widgets/jupyterlab-manager` for ipywidgets). Versions of `@jupyterlab` packages can be found in the commits for the release tag at https://github.com/jupyterlab/jupyterlab/releases
```

Comment:
There is a federated extension for ipywidgets now: https://pypi.org/project/jupyterlab-widgets/.

Reply:
> There is a federated extension for ipywidgets now: https://pypi.org/project/jupyterlab-widgets/.

thanks for bringing this up. I am not sure if we can integrate ipywidgets to desktop app using that python package. For the desktop app, all js front-end components are bundled with the electron-app and the conda environment acts as the backend/server of the desktop app. that's why everything on the front-end needs to be pre-bundled using webpack (https://github.com/jupyterlab/jupyterlab_app/blob/master/webpack.browser.js and https://github.com/jupyterlab/jupyterlab_app/blob/master/src/browser/extensions/index.ts). Please let me know if there is a better way to bundle the ipywidgets js package.

Reply:
I think compiling in ipywidgets to start with is great (and a sign that leaving in the ability to compile things into jlab was important).

It would be interesting to explore how you might be able to take advantage of prebuilt extensions. Presumably your core webpack bundle can load prebuilt extension bundles just as well as our core webpack bundle.

Reply:
Thank you @jasongrout . I will be researching that. We will need it to support jlab extensions in the App as well.